src/overlay/TCPPeer.cpp

```diff
@@ -588,30 +625,33 @@ TCPPeer::startRead()
         // If we get here it's because we broke out of the input loop above
         // early (via shouldYield) which means it's time to bounce off a the
         // per-peer scheduler queue to throttle further input.
+        // Note: this can only happen on the main thread, because background
+        // thread is outside of main thread Scheduler's discipline
+        releaseAssert(threadIsMain());
         scheduleRead();
     }
 }
 
 size_t
 TCPPeer::getIncomingMsgLength()
 {
-    size_t length = static_cast<size_t>(mIncomingHeader[0]);
+    std::lock_guard<std::recursive_mutex> guard(mStateMutex);
+    size_t length = static_cast<size_t>(mThreadVars.getIncomingHeader()[0]);
     length &= 0x7f; // clear the XDR 'continuation' bit
     length <<= 8;
-    length |= mIncomingHeader[1];
+    length |= mThreadVars.getIncomingHeader()[1];
     length <<= 8;
-    length |= mIncomingHeader[2];
+    length |= mThreadVars.getIncomingHeader()[2];
     length <<= 8;
-    length |= mIncomingHeader[3];
+    length |= mThreadVars.getIncomingHeader()[3];
```

Comment:
I know this was here before, but these unchecked accesses are little sketchy. In practice `mThreadVars.getIncomingHeader()` is always resized to be at least `4` elements large before calling this function, but it seems a bit error prone to me. What do you think of adding an assert that the vector is large enough? Or if there are performance concerns there, at least a comment in the header file stating that `mThreadVars.getIncomingHeader.size()` must be at least `4`?

Reply:
Done!